libs/auth/src/inserted_smart_card_auth_with_memory_card.ts

```diff
@@ -0,0 +1,294 @@
+  Result,
+  throwIllegalValue,
+  wrapException,
+} from '@votingworks/basics';
+import {
+  Card,
+  CardSummary,
+  ElectionDefinition,
+  getBallotStyle,
+  getPrecinctById,
+  InsertedSmartCardAuth,
+  Optional,
+  PrecinctSelection,
+  User,
+} from '@votingworks/types';
+import { utcTimestamp } from '@votingworks/utils';
+
+import {
+  InsertedSmartCardAuthApi,
+  InsertedSmartCardAuthConfig,
+} from './inserted_smart_card_auth';
+import {
+  CARD_POLLING_INTERVAL_MS,
+  parseUserFromCardSummary,
+} from './memory_card';
+
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-ignore
+global.fetch = fetch;
+
+const VOTER_CARD_EXPIRATION_SECONDS = 60 * 60; // 1 hour
+
+type AuthAction =
+  | { type: 'read_card'; cardSummary: CardSummary }
+  | { type: 'check_pin'; pin: string };
+
+/**
+ * An implementation of the dipped smart card auth API, backed by a memory card
+ *
+ * Since this is just a stopgap until we implement InsertedSmartCardAuthWithJavaCard, I haven't
+ * implemented the following:
+ * - Locking to avoid concurrent card writes
+ * - Logging
+ * - Tests
+ */
+export class InsertedSmartCardAuthWithMemoryCard
+  implements InsertedSmartCardAuthApi
+{
+  private authStatus: InsertedSmartCardAuth.AuthStatus;
+  private readonly card: Card;
+  private config: InsertedSmartCardAuthConfig;
+
+  constructor({
+    card,
+    config,
+  }: {
+    card: Card;
+    config: Omit<
+      InsertedSmartCardAuthConfig,
+      'electionDefinition' | 'precinctSelection'
+    >;
+  }) {
+    this.authStatus = InsertedSmartCardAuth.DEFAULT_AUTH_STATUS;
+    this.card = card;
+    this.config = config;
+
+    setInterval(
+      async () => {
+        try {
+          const newCardSummary = await this.card.readSummary();
+          this.updateAuthStatus({
+            type: 'read_card',
+            cardSummary: newCardSummary,
+          });
+        } catch {
+          // Swallow errors so that they don't crash the auth instance and containing backend
+        }
+      },
+      CARD_POLLING_INTERVAL_MS,
+      true
+    );
+  }
+
+  getAuthStatus(): InsertedSmartCardAuth.AuthStatus {
+    return this.authStatus;
+  }
+
+  checkPin({ pin }: { pin: string }): void {
+    this.updateAuthStatus({ type: 'check_pin', pin });
+  }
+
+  readCardData<T>({
+    schema,
+  }: {
+    schema: z.ZodSchema<T>;
+  }): Promise<Result<Optional<T>, SyntaxError | z.ZodError | Error>> {
+    return this.card.readLongObject(schema);
+  }
+
+  async writeCardData<T>({
+    data,
+    schema,
+  }: {
+    data: T;
+    schema: z.ZodSchema<T>;
+  }): Promise<Result<void, Error>> {
+    try {
+      await this.card.writeLongObject(data);
+    } catch (error) {
+      return wrapException(error);
+    }
+
+    // Verify that the write was in fact successful by reading the data
+    const readResult = await this.readCardData({ schema });
+    if (readResult.isErr()) {
+      return err(new Error('Verification of write by reading data failed'));
+    }
```

Comment:
We used to perform this validation client-side, e.g.

https://github.com/votingworks/vxsuite/blob/46b0c7ec6da8b39f46eec25b8133e48272cec097/apps/vx-scan/frontend/src/screens/poll_worker_screen.tsx#L79-L90

Especially with the shift to react-query and hooks on the frontend, I found it cleaner to encapsulate this validation on the backend.

I do recognize that I'm not validating the `timeSaved` field as we used to on the frontend. Still getting a sense of how important that is before adding generalized support for field validation on the backend. My hunch is that this will all be less relevant with the shift to Java Cards, which provide more storage space than our current memory cards. The main silent error today seems to be data truncation, hence this logic on the frontend:

https://github.com/votingworks/vxsuite/blob/46b0c7ec6da8b39f46eec25b8133e48272cec097/apps/vx-scan/frontend/src/screens/poll_worker_screen.tsx#L300-L311